src/frequenz/dispatch/_actor_dispatcher.py

```diff
@@ -295,13 +296,16 @@ async def _run(self) -> None:
             async for dispatch in self._dispatch_rx:
                 await self._handle_dispatch(dispatch)
         else:
-            retry_recv = self._retrier.new_receiver()
-
-            async for selected in select(retry_recv, self._dispatch_rx):
-                if retry_recv.triggered(selected):
-                    self._retrier.retry(selected.message)
-                elif self._dispatch_rx.triggered(selected):
-                    await self._handle_dispatch(selected.message)
+            try:
+                retry_recv = self._retrier.new_receiver()
+
+                async for selected in select(retry_recv, self._dispatch_rx):
+                    if retry_recv.triggered(selected):
+                        self._retrier.retry(selected.message)
+                    elif self._dispatch_rx.triggered(selected):
+                        await self._handle_dispatch(selected.message)
+            finally:
+                self._retrier.cleanup()
```

Comment:
If this is a background service, you can even instantiate the retrier here inside an `async with` block (you could still keep it as an instance attribute if you need to use it in many other places, but if not I think it makes sense to instantiate it in the `run`:

```py
async def _run(self) -> None:
        """Run the background service."""
        async with RetryFailedDispatches(...) as retrier:
            ...
```